src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -28,69 +29,489 @@
+  }
+
+  inline void clear_bit(ssize_t idx) {
+    assert((idx >= 0) && (idx < _num_bits), "precondition");
+    assert(idx >= 0, "precondition");
+    size_t array_idx = idx >> LogBitsPerWord;
+    uintx bit_number = idx & right_n_bits(LogBitsPerWord);
+    uintx the_bit = nth_bit(bit_number);
+    _bitmap[array_idx] &= ~the_bit;
+  }
+
+  inline bool is_set(ssize_t idx) const {
+    assert((idx >= 0) && (idx < _num_bits), "precondition");
+    assert(idx >= 0, "precondition");
+    size_t array_idx = idx >> LogBitsPerWord;
+    uintx bit_number = idx & right_n_bits(LogBitsPerWord);
+    uintx the_bit = nth_bit(bit_number);
+    return (_bitmap[array_idx] & the_bit)? true: false;
+  }
+
+  // Return the index of the first set bit which is greater or equal to start_idx.  If not found, return _num_bits.
+  inline ssize_t find_next_set_bit(ssize_t start_idx) const;
+
+  // Return the index of the first set bit which is greater or equal to start_idx and less than boundary_idx.
+  // If not found, return boundary_idx
+  inline ssize_t find_next_set_bit(ssize_t start_idx, ssize_t boundary_idx) const;
+
+  // Return the index of the last set bit which is less or equal to start_idx.  If not found, return -1.
+  inline ssize_t find_prev_set_bit(ssize_t last_idx) const;
+
+  // Return the index of the last set bit which is less or equal to start_idx and greater than boundary_idx.
+  // If not found, return boundary_idx.
+  inline ssize_t find_prev_set_bit(ssize_t last_idx, ssize_t boundary_idx) const;
+
+  // Return the smallest index at which a run of num_bits consecutive ones is found, where return value is >= start_idx
+  // and return value < _num_bits.  If no run of num_bits consecutive ones is found within the target range, return _num_bits.
+  inline ssize_t find_next_consecutive_bits(size_t num_bits, ssize_t start_idx) const;
+
+  // Return the smallest index at which a run of num_bits consecutive ones is found, where return value is >= start_idx
+  // and return value < boundary_idx.  If no run of num_bits consecutive ones is found within the target range,
+  // return boundary_idx.
+  ssize_t find_next_consecutive_bits(size_t num_bits, ssize_t start_idx, ssize_t boundary_idx) const;
+
+  // Return the largest index at which a run of num_bits consecutive ones is found, where return value is <= last_idx and > -1.
+  // If no run of num_bits consecutive ones is found within the target range, return -1.
+  inline ssize_t find_prev_consecutive_bits(size_t num_bits, ssize_t last_idx) const;
+
+  // Return the largest index at which a run of num_bits consecutive ones is found, where return value is <= last_idx and > -1.
+  // If no run of num_bits consecutive ones is found within the target range, return -1.
+  ssize_t find_prev_consecutive_bits(size_t num_bits, ssize_t last_idx, ssize_t boundary_idx) const;
+};
+
+// Each ShenandoahHeapRegion is associated with a ShenandoahFreeSetPartitionId.
+enum ShenandoahFreeSetPartitionId : uint8_t {
```

Comment:
Make this an 'enum class' instead.
I see that you are using the enum as index, in this case you can declare the enum class like 'enum class ShenandoahFreeSetPartitionId : uint8_t'

Reply:
Thanks for this suggestion.  Done.

Reply:
Thanks.  This is done.